index.js

```diff
@@ -3700,6 +3700,60 @@
   S.pairs =
   def('pairs', {}, [$.StrMap(a), $.Array($.Pair($.String, a))], pairs);
 
+  //# insert :: String -> a -> StrMap a -> StrMap a
+  //.
+  //. Inserts a given key and value into a shallow clone of a given
+  //. string map. Overrides the value if a key already exists.
+  //.
+  //. Equivalent to Haskell's `insert` function. Similar to Clojure's `assoc`
+  //. functiom.
+  //.
+  //. ```javascript
+  //. > S.insert('c', 3, {a: 1, b: 2})
+  //. {a: 1, b: 2, c: 3}
+  //.
+  //. > S.insert('a', 4, {a: 1, b: 2})
+  //. {a: 4, b: 2}
+  //. ```
+  function insert(key, val, strMap) {
+    var result = Z.concat(strMap, {});
+    result[key] = val;
+    return result;
+  }
+  S.insert =
+  def('insert', {}, [$.String, a, $.StrMap(a), $.StrMap(a)], insert);
+
+  //# remove :: String -> StrMap a -> StrMap a
+  //.
+  //. Removes a given key and its value from a shallow clone of a given
+  //. string map. The original string map is returned if the key doesn't exist.
+  //.
+  //. Equivalent to Haskell's `delete` function. Similar to Clojure's `dissoc`
+  //. functiom.
+  //.
+  //. ```javascript
+  //. > S.remove('c', {a: 1, b: 2, c: 3})
+  //. {a: 1, b: 2}
+  //.
+  //. > S.remove('c', {a: 1, b: 2})
+  //. {a: 1, b: 2}
+  //.
+  //. > S.remove('c', {})
+  //. {}
+  //. ```
+  function remove(key, strMap) {
+    if (key in strMap) {
```

Comment:
This is unsafe due to inherited properties. `'toString' in {}`, for example, evaluates to `true`.

Is this implementation suitable?

```javascript
var result = Z.concat(strMap, {});
delete result[key];
return result;
```

Reply:
Wasn't aware of the inherited properties issue. The initial thought was to use `concat` in the way you wrote. But it bothered me to add a property to an object and then delete it. In addition, I liked the efficiency of Haskell's `delete` function where it returns the original `Map` if the key didn't exist.

What about `if (strMap.hasOwnProperty(key))`?

Reply:
> What about `if (strMap.hasOwnProperty(key))`?

This would get us a bit closer, but we must allow for the possibility of `'hasOwnProperty'` being one of the keys. We'd need `Object.prototype.hasOwnProperty.call(strMap, key)`.

I think we should match the semantics of `Z.concat`, though, which is interested in *enumerable* properties rather than *own* properties. For example:

```javascript
function Point() {}
Point.prototype.x = 0;
Point.prototype.y = 0;

Z.concat(new Point(), {});  // => {x: 0, y: 0}
```

Reply:
Good catch on the `'hasOwnProperty'` as a key issue.

I understand `Z.concat` uses for..in to iterate through keys of enumerable and inherited enumerable properties. We could use`Object.prototype.propertyIsEnumerable.call(strMap, key)`, but that will only tell us if a property is enumerable. It will return `false` if a property is _inherited_ enumerable as in your `Point` example above. So, if we want to stick with returning the original string map when it doesn't contain the key to remove, we need to iterate through the keys before determining whether we can simply return the original string map:

```javascript
  function remove(key, strMap) {
    var found = false;
    var result = {};
    for (var k in strMap) {
      if (k === key) {
        found = true;
      } else {
        result[k] = strMap[k];
      }
    }
    if (found) {
      return result;
    } else {
      return strMap;
    }
  }
```
If optimizing on removing non-existing keys more often, we could search for the key first before copying.

Reply:
> ```javascript
> if (found) {
>   return result;
> } else {
>   return strMap;
> }
> ```

If we're to do the work to generate `result`, why not return it in all cases? Sanctuary users, ideally, will not care about object identity, so it shouldn't be necessary to have `S.remove(key, strMap) === strMap` evaluate to `true` in certain cases. It may, though, be useful to guarantee that this expression *never* evaluates to `true` (the `delete`-based implementation above takes advantage of the fact that `Z.concat` always returns a new string map).

Reply:
The idea was it was more efficient to return the original string map back if the key didn't exist. If we're going to build the `result` object anyway, the performance/efficiency benefit is diminished. I can imagine there are still benefits as the temporary `result` will get garbage collected. But I don't know how much that matters vs the benefit of guaranteeing `remove` never returns the original string map.

I can use your `Z.concat` implementation or I can use the above implementation without the code to return the original string map if the key wasn't found.

Reply:
I appreciate @Avaq's countervailing input on performance matters, as I'm inclined to place more weight on consistency (always returning a new string map, in this case) than on performance. ;)

I'd like to know what others think. @Avaq? @gabejohnson? @safareli?

Reply:
If we know that user will not mutate input of the function, than it is fine to return same value back if key is not there. But problem is that you don't know it and i'm not sure if you can assume it, some people might use sanctuary just for type checking not for controlling effects.

but if input `isFrozen` then we know that it will not be mutated and it might be correct to return same value.

Reply:
I don't know that we necessarily _should_ match the semantics of `S.concat` as it works on objects in general. I think we may inadvertantly cause confusion by not explicitly defining `StrMap`. In my opinion inherited properties should not be copied for a `StrMap`.

Reply:
I don't know that we necessarily _should_ match the semantics of `S.concat` as it works on objects in general. I think we may inadvertantly cause confusion by not explicitly defining `StrMap`. In my opinion inherited properties should not be copied for a `StrMap`.

Reply:
@gabejohnson You might want to join the discussion here: https://github.com/sanctuary-js/sanctuary-def/issues/164

On this specific case, it looks like the whole `result` object must be computed either way, and the choice is between returning a clone or the original. My preference would be returning a clone. I think the argument of memory usage does not outweigh the argument of consistency. In fact, if this function is used in a pure pipeline, the original object will be garbage collected anyway. If this function is used in a way where references are kept to *both* the input and output, I think users would be safer with the output being a clone. I'm in favor of `return result;`.

Reply:
Thanks for raising your concern, @gabejohnson. Let's discuss the matter in sanctuary-js/sanctuary-type-classes#65.

> I think we should match the semantics of `Z.concat`, though, which is interested in *enumerable* properties rather than *own* properties.

We may to decide to change the behaviour of `Z.concat`, but I stand by the first clause of the sentence above. The only way to be sure that we're matching the semantics of `Z.concat` is to implement `S.insert` and `S.remove` in terms of `Z.concat`. Let's use the implementation I proposed at the beginning of this thread. If the behaviour of `Z.concat` changes in the future the behaviour of `S.insert` and `S.remove` will adjust accordingly.

If you're concerned, @samueller, that a future release of sanctuary-type-classes could introduce a subtle regression, you're welcome to include tests for the handling of inherited properties. If and when we were to upgrade to a version of sanctuary-type-classes with different semantics we would then update the assertions. These tests would serve to remind us to mention breaking changes to `S.insert` and `S.remove` in future release notes rather than to bind us to the current semantics.

Reply:
So, which implementation?
```javascript
var result = Z.concat(strMap, {});
delete result[key];
return result;
```
or
```javascript
var result = {};
for (var k in strMap) {
  if (k !== key) result[k] = strMap[k];
}
return result;
```
First one is automatically updated when/if `Z.concat` is updated. Second one is more efficient assuming `delete` is more expensive than the `n` key comparisons.

Reply:
So, which implementation?
```javascript
var result = Z.concat(strMap, {});
delete result[key];
return result;
```
or
```javascript
var result = {};
for (var k in strMap) {
  if (k !== key) result[k] = strMap[k];
}
return result;
```
First one is automatically updated when/if `Z.concat` is updated. Second one is more efficient assuming `delete` is more expensive than the `n` key comparisons.

Tests for inherited properties make sense.

Reply:
Let's use this implementation:

```javascript
var result = Z.concat(strMap, {});
delete result[key];
return result;
```

> [The implementation above] is automatically updated when/if `Z.concat` is updated.

I consider this a requirement. I'm keen to avoid inconsistencies such as the one documented in ramda/ramda#1941.

Reply:
We also have `S.keys`. Whichever way we go with properties to include, I think we should have (pseudo code, I think I get to the idea):

```js
function insert_test(key, val, strMap) {
  new Set(S.keys(strMap).concat([key])) === new Set(S.keys(S.insert(key, val, strMap)))
}
```
?

Reply:
Great point, Kevin! It would be good to test this property. Here are a few more:

  - `S.insert (key) (val) (S.insert (key) (val) (map)) = S.insert (key) (val) (map)`
  - `S.remove (key) (S.remove (key) (map)) = S.remove (key) (map)`
  - `S.sort (S.keys (map)) = S.sort (S.keys (S.map (S.I) (map)))`

Reply:
I like the tests and have added them, except the last assertion you mentioned @davidchambers. I assume that test was meant for the `keys` or the `map` function? I can add that to this pull request if you want.

Reply:
> I assume that test was meant for the `keys` or the `map` function?

Yes, you're right. It would assert that `S.map` does not "promote" any properties which `S.keys` does not see. Let's leave that for a separate pull request.

I look forward to seeing your changes. Feel free to push the changes you have made so far even if there are more to come. You can tidy the commit history before we merge the pull request. :)